src/shims/unix/socket.rs

```diff
@@ -93,6 +153,8 @@ impl FileDescription for SocketPair {
         // Do full read / partial read based on the space available.
         // Conveniently, `read` exists on `VecDeque` and has exactly the desired behavior.
         let actual_read_size = readbuf.buf.read(bytes).unwrap();
+        // Set event mask.
+        self.check_and_update_readiness(ecx).unwrap();
```

Comment:
like with my other comment on the other file descriptor, the `check_readiness` method seems a bit odd, as it checks all the flags. It's probably a fairly robust way to implement all this, but it does seem odd and needs some explanation.

Reply:
Yup, the previous implementation of ``check_readiness`` and ``update_readiness`` implementation is incorrect. What the current version do is:
1. If there is any event (read/write/close), we check the readiness of ALL supported flags for that file description. 
2. At ``update_readiness``, we only return notification if the there is a match between ``ready flag for file description`` and ``flag that required notification for epoll_event``.

Let's say we have ``epoll_event`` that registered ``EPOLLIN``, and the file description is ready for both ``EPOLLIN`` and ``EPOLLOUT``, what will happen is:
1. ``check_readiness`` will return a vector that contains both ``EPOLLIN`` and ``EPOLLOUT``
2. ``update_readiness`` will only return notification for ``EPOLLIN`` since the notification is only requested for ``EPOLLIN``. 

Even if the ``check_readiness`` is invoked from ``write``, we still need to check readiness of ALL supported flags the reflect the current state of file description, we can't just update the ``EPOLLIN`` flag alone. 

Here is a related SO post on why all flags should be checked:
> To be perhaps clearer, epoll_wait() won't return an fd unless something changed on that socket, but if something did change, it returns all the flags representing the current state.
- https://stackoverflow.com/questions/59288735/epoll-wait-return-epollout-even-with-epollet-flag